File: app/src/main/java/com/zebra/barcodeintellgencetools/ItemDetailFragment.java
package com.zebra.barcodeintellgencetools;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.fragment.app.Fragment;

import com.google.android.material.appbar.CollapsingToolbarLayout;
import com.zebra.barcodeintellgencetools.api.APIContent;
import com.zebra.barcodeintellgencetools.api.RetrieveAPITask;
import com.zebra.savanna.BaseAPI;
import com.zebra.savanna.Symbology;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;

/**
 * A fragment representing a single Item detail screen.
 * This fragment is either contained in a {@link ItemListActivity}
 * in two-pane mode (on tablets) or a {@link ItemDetailActivity}
 * on handsets.
 */
public class ItemDetailFragment extends Fragment implements View.OnClickListener {
    /**
     * The fragment argument representing the item ID that this fragment
     * represents.
     */
    static final String ARG_ITEM_ID = "item_id";
    public static ItemDetailFragment Instance;
    /**
     * The item content this fragment is presenting.
     */
    private static APIContent.ApiItem mItem;
    private static String details = "";
    private static Bitmap barcodeImage;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public ItemDetailFragment() {
    }

    public void onPostExecute(Object apiData) {
        ViewGroup root = (ViewGroup) getView();
        if (root == null) return;
        if (apiData instanceof byte[]) {
            ImageView barcode = root.findViewById(R.id.barcode);
            byte[] data = (byte[]) apiData;
            barcodeImage = BitmapFactory.decodeByteArray(data, 0, data.length);
            barcode.setImageBitmap(barcodeImage);
        } else {
            JSONObject json = (JSONObject) apiData;
            TextView results = root.findViewById(R.id.resultData);
            if (results != null) {
                try {
                    if (details.equals("")) {
                        details = json.toString(2);
                    } else
                        details += "\n" + json.toString(2);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                results.setText(details);
            }
        }
    }

    void routeScanData(String barcode, String symbology) {
        ViewGroup root = (ViewGroup) getView();
        if (root == null) return;
        TextView results = root.findViewById(R.id.resultData);
        symbology = symbology.substring("label-type-".length());
        if (symbology.equals("upce0")){
            symbology = "upce";
            if (barcode.length() == 6)
                barcode = "0" + barcode + "0";
        }
        switch (mItem.id) {
            case "1":
                EditText barcodeText = root.findViewById(R.id.barcodeText);
                barcodeText.setText(barcode);

                Spinner barcodeType = root.findViewById(R.id.barcodeTypes);
                int index = Arrays.asList(Symbology.values()).indexOf(Symbology.fromString(symbology));
                if (index > -1)
                    barcodeType.setSelection(index);
                return;
            case "2":
                details = "";
                results.setText(details);
                new RetrieveAPITask().execute("foodUpc", barcode);
                new RetrieveAPITask().execute("drugUpc", barcode);
                return;
            case "3":
                details = "";
                results.setText(details);
                new RetrieveAPITask().execute("lookup", barcode);
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Instance = this;
        Bundle args = getArguments();
        if (args != null && args.containsKey(ARG_ITEM_ID)) {
            String key = args.getString(ARG_ITEM_ID);

            if (mItem != null && key != null && !key.equals(mItem.id)) {
                barcodeImage = null;
                details = "";
            }

            // Load the item content specified by the fragment
            // arguments. In a real-world scenario, use a Loader
            // to load content from a content provider.
            mItem = APIContent.ITEM_MAP.get(key);

            Activity activity = this.getActivity();
            CollapsingToolbarLayout appBarLayout = activity == null ? null : (CollapsingToolbarLayout) activity.findViewById(R.id.toolbar_layout);
            if (appBarLayout != null) {
                appBarLayout.setTitle(mItem.content);
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        ViewGroup root = (ViewGroup) inflater.inflate(R.layout.item_detail, container, false);

        SharedPreferences sharedPreferences =
                PreferenceManager.getDefaultSharedPreferences(getContext());
        BaseAPI.APIKey = sharedPreferences.getString("apikey", "");
        System.out.println(BaseAPI.APIKey);
        // Show the item content as text in a TextView.
        if (mItem != null) {
            ((TextView) root.findViewById(R.id.item_detail)).setText(mItem.details);

            switch (mItem.id) {
                case "1":
                    View createView = inflater.inflate(R.layout.create_barcode, container, false);
                    Spinner types = createView.findViewById(R.id.barcodeTypes);
                    Context context = getContext();
                    if (context != null)
                        types.setAdapter(new ArrayAdapter<>(context, android.R.layout.simple_spinner_dropdown_item, Symbology.values()));
                    Button create = createView.findViewById(R.id.createBarcode);
                    create.setOnClickListener(this);
                    if (barcodeImage != null) {
                        ImageView barcode = createView.findViewById(R.id.barcode);
                        barcode.setImageBitmap(barcodeImage);
                    }
                    root.addView(createView);
                    break;
                case "2":
                    View recallView = inflater.inflate(R.layout.fda_recall, container, false);
                    Button recalls = recallView.findViewById(R.id.fdaSearch);
                    recalls.setOnClickListener(this);
                    TextView recallResults = recallView.findViewById(R.id.resultData);
                    if (details.equals(""))
                        recallResults.setText(R.string.fda_scan);
                    else
                        recallResults.setText(details);
                    root.addView(recallView);
                    break;
                case "3":
                    View lookupView = inflater.inflate(R.layout.upc_lookup, container, false);
                    TextView results = lookupView.findViewById(R.id.resultData);
                    if (details.equals(""))
                        results.setText(R.string.scan);
                    else
                        results.setText(details);
                    root.addView(lookupView);
                    break;
            }
        }

        return root;
    }

    @Override
    public void onClick(View v) {
        ViewGroup root = (ViewGroup) this.getView();
        if (root == null) return;
        TextView results = root.findViewById(R.id.resultData);
        switch (mItem.id) {
            case "1":
                EditText barcodeText = root.findViewById(R.id.barcodeText);
                Spinner barcodeType = root.findViewById(R.id.barcodeTypes);

                new RetrieveAPITask().execute("create", barcodeText.getText().toString(), barcodeType.getSelectedItem().toString());
                return;
            case "2":
                details = "";
                results.setText(details);
                EditText searchText = root.findViewById(R.id.fdaSearchTerm);
                new RetrieveAPITask().execute("deviceSearch", searchText.getText().toString());
                new RetrieveAPITask().execute("drugSearch", searchText.getText().toString());
        }
    }
}

File: BarcodeIntelligenceTools/ItemDetailFragment.cs
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Preferences;
using Android.Views;
using Android.Widget;
using AndroidX.Fragment.App;
using BarcodeIntelligenceTools.API;
using Google.Android.Material.AppBar;
using Savanna;
using Savanna.Models;
using System;
using System.Linq;

namespace BarcodeIntelligenceTools
{
    /// <summary>
    /// A fragment representing a single Item detail screen.
    /// This fragment is either contained in a <see cref="ItemListActivity"/>
    /// in two-pane mode (on tablets) or a <see cref="ItemDetailActivity"/>
    /// on handsets.
    /// </summary>
    public class ItemDetailFragment : Fragment, View.IOnClickListener
    {
        /// <summary>
        /// The fragment argument representing the item ID that this fragment
        /// represents.
        /// </summary>
        public const string ArgItemId = "item_id";
        public static ItemDetailFragment Instance { get; set; }
        /// <summary>
        /// The item content this fragment is presenting.
        /// </summary>
        private static ApiItem _item;
        private static string _details = "";
        private static Bitmap _barcodeImage;

        /// <summary>
        /// Mandatory empty constructor for the fragment manager to instantiate the
        /// fragment (e.g. upon screen orientation changes).
        /// </summary>
        public ItemDetailFragment()
        {
        }

        private void OnPostExecute(object apiData)
        {
            ViewGroup root = (ViewGroup)View;
            if (root == null) return;
            ImageView barcode = root.FindViewById<ImageView>(Resource.Id.barcode);
            TextView results = root.FindViewById<TextView>(Resource.Id.resultData);
            if (apiData is byte[] data)
            {
                _barcodeImage = BitmapFactory.DecodeByteArray(data, 0, data.Length);
                barcode.SetImageBitmap(_barcodeImage);
                barcode.Visibility = ViewStates.Visible;
                results.Visibility = ViewStates.Gone;
            }
            else if (apiData is Error<string> e)
            {
                OnPostExecute(e.DeveloperMessage == null || e.DeveloperMessage == e.Message ? e.Message : $"{e.Message}: {e.DeveloperMessage}");
            }
            else if (apiData is Error<DeveloperMessage> dm)
            {
                OnPostExecute(dm.DeveloperMessage == null || dm.DeveloperMessage.Fault.FaultString == dm.Message ? dm.Message : $"{dm.Message}: {dm.DeveloperMessage.Fault.FaultString}");
            }
            else if (apiData is Exception ex)
            {
                OnPostExecute(ex.Message);
            }
            else
            {
                string json = (string)apiData;
                results.Visibility = ViewStates.Visible;
                if (_details.Length == 0)
                {
                    _details = json;
                }
                else
                {
                    _details += "\n" + json;
                }
                results.Text = _details;
                if (barcode != null)
                    barcode.Visibility = ViewStates.Gone;
            }
        }

        public async void RouteScanData(string barcode, string symbology)
        {
            if (!(View is ViewGroup root)) return;
            TextView results = root.FindViewById<TextView>(Resource.Id.resultData);
            symbology = symbology.Substring("label-type-".Length);
            if (symbology.Equals("upce0"))
            {
                symbology = "upce";
                if (barcode.Length == 6)
                    barcode = "0" + barcode + "0";
            }
            switch (_item.Id)
            {
                case "1":
                    EditText barcodeText = root.FindViewById<EditText>(Resource.Id.barcodeText);
                    barcodeText.Text = barcode;

                    Spinner barcodeType = root.FindViewById<Spinner>(Resource.Id.barcodeTypes);
                    int index = Array.IndexOf(Enum.GetValues(typeof(Symbology)), Enum.Parse<Symbology>(symbology.Replace('-', '_')));
                    if (index > -1)
                        barcodeType.SetSelection(index);
                    return;
                case "2":
                    _details = "";
                    results.Text = _details;
                    try
                    {
                        OnPostExecute(await FDARecall.FoodUpcAsync(barcode));
                    }
                    catch (Exception e)
                    {
                        OnPostExecute(e);
                    }
                    try
                    {
                        OnPostExecute(await FDARecall.DrugUpcAsync(barcode));
                    }
                    catch (Exception e)
                    {
                        OnPostExecute(e);
                    }
                    return;
                case "3":
                    _details = "";
                    results.Text = _details;
                    EditText upc = root.FindViewById<EditText>(Resource.Id.upc);
                    upc.Text = barcode;
                    try
                    {
                        OnPostExecute(await UPCLookup.LookupAsync(barcode));
                    }
                    catch (Exception e)
                    {
                        OnPostExecute(e);
                    }
                    return;
            }
        }


        public override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Instance = this;
            Bundle args = Arguments;
            if (args != null && args.ContainsKey(ArgItemId))
            {
                string key = args.GetString(ArgItemId);

                if (_item != null && key != null && !key.Equals(_item.Id))
                {
                    _barcodeImage = null;
                    _details = "";
                }

                // Load the item content specified by the fragment
                // arguments. In a real-world scenario, use a Loader
                // to load content from a content provider.
                _item = APIContent.Items[key].Item;

                var activity = Activity;
                var appBarLayout = activity?.FindViewById<CollapsingToolbarLayout>(Resource.Id.toolbar_layout);
                if (appBarLayout != null)
                {
                    appBarLayout.SetTitle(_item.Content);
                }
            }
        }

        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
        {
            var root = (ViewGroup)inflater.Inflate(Resource.Layout.item_detail, container, false);
            var sharedPreferences = PreferenceManager.GetDefaultSharedPreferences(Context);
            BaseAPI.APIKey = sharedPreferences.GetString("apikey", "");
            Console.WriteLine(BaseAPI.APIKey);
            // Show the item content as text in a TextView.
            if (_item != null)
            {
                root.FindViewById<TextView>(Resource.Id.item_detail).Text = _item.Details;

                switch (_item.Id)
                {
                    case "1":
                        View createView = inflater.Inflate(Resource.Layout.create_barcode, container, false);

                        TextView createResults = createView.FindViewById<TextView>(Resource.Id.resultData);
                        createResults.Text = _details;

                        Button create = createView.FindViewById<Button>(Resource.Id.createBarcode);
                        create.SetOnClickListener(this);

                        Spinner types = createView.FindViewById<Spinner>(Resource.Id.barcodeTypes);
                        Context context = Context;
                        if (context != null)
                        {
                            Array syms = Enum.GetValues(typeof(Symbology));
                            Symbology[] values = new Symbology[syms.Length];
                            syms.CopyTo(values, 0);
                            types.Adapter = new ArrayAdapter(context, Android.Resource.Layout.SimpleSpinnerDropDownItem, values.Select(s => s.ToString().Replace('_', '-')).ToList());
                        }
                        ImageView barcode = createView.FindViewById<ImageView>(Resource.Id.barcode);
                        if (_barcodeImage == null && _details.Equals(""))
                        {
                            barcode.Visibility = ViewStates.Gone;
                            createResults.Visibility = ViewStates.Visible;
                        }
                        else
                        {
                            barcode.SetImageBitmap(_barcodeImage);
                            barcode.Visibility = ViewStates.Visible;
                            createResults.Visibility = ViewStates.Gone;
                        }
                        root.AddView(createView);
                        break;
                    case "2":
                        View recallView = inflater.Inflate(Resource.Layout.fda_recall, container, false);
                        Button recalls = recallView.FindViewById<Button>(Resource.Id.fdaSearch);
                        recalls.SetOnClickListener(this);
                        TextView recallResults = recallView.FindViewById<TextView>(Resource.Id.resultData);
                        recallResults.Text = _details;
                        root.AddView(recallView);
                        break;
                    case "3":
                        View lookupView = inflater.Inflate(Resource.Layout.upc_lookup, container, false);
                        TextView results = lookupView.FindViewById<TextView>(Resource.Id.resultData);
                        results.Text = _details;

                        Button lookup = lookupView.FindViewById<Button>(Resource.Id.upc_lookup);
                        lookup.SetOnClickListener(this);

                        root.AddView(lookupView);
                        break;
                }
            }

            return root;
        }

        public async void OnClick(View v)
        {
            if (!(View is ViewGroup root)) return;

            TextView results = root.FindViewById<TextView>(Resource.Id.resultData);
            _details = "";
            results.Text = _details;
            try
            {
                switch (_item.Id)
                {
                    case "1":
                        EditText barcodeText = root.FindViewById<EditText>(Resource.Id.barcodeText);
                        Spinner barcodeType = root.FindViewById<Spinner>(Resource.Id.barcodeTypes);
                        var symbology = Enum.Parse<Symbology>(barcodeType.SelectedItem.ToString().Replace('-', '_'));
                        OnPostExecute(await CreateBarcode.CreateAsync(symbology, barcodeText.Text, ItemListActivity.Density, Rotation.Normal, true));
                        return;
                    case "2":
                        EditText searchText = root.FindViewById<EditText>(Resource.Id.fdaSearchTerm);
                        try
                        {
                            OnPostExecute(await FDARecall.DeviceSearchAsync(searchText.Text));
                        }
                        catch (Exception e)
                        {
                            OnPostExecute(e);
                        }
                        OnPostExecute(await FDARecall.DrugSearchAsync(searchText.Text));
                        return;
                    case "3":
                        EditText lookupText = root.FindViewById<EditText>(Resource.Id.upc);
                        OnPostExecute(await UPCLookup.LookupAsync(lookupText.Text));
                        return;
                }
            }
            catch (Exception e)
            {
                OnPostExecute(e);
            }
        }
    }
}